contracts/cw-governance/src/contract.rs

```diff
@@ -0,0 +1,359 @@
+use std::usize;
+
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    to_binary, Addr, Binary, CosmosMsg, Deps, DepsMut, Empty, Env, MessageInfo, Reply, Response,
+    StdResult, SubMsg,
+};
+use cw2::{get_contract_version, set_contract_version};
+use cw_storage_plus::Bound;
+use cw_utils::parse_reply_instantiate_data;
+
+use cw_governance_interface::voting;
+
+use crate::error::ContractError;
+use crate::msg::{ExecuteMsg, InstantiateMsg, ModuleInstantiateInfo, QueryMsg};
+use crate::query::{DumpStateResponse, GetItemResponse};
+use crate::state::{
+    Config, CONFIG, GOVERNANCE_MODULES, GOVERNANCE_MODULE_COUNT, ITEMS, VOTING_MODULE,
+};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:cw-governance";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+const GOV_MODULE_REPLY_ID: u64 = 0;
+const VOTE_MODULE_INSTANTIATE_REPLY_ID: u64 = 1;
+const VOTE_MODULE_UPDATE_REPLY_ID: u64 = 2;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let config = Config {
+        name: msg.name,
+        description: msg.description,
+        image_url: msg.image_url,
+    };
+    CONFIG.save(deps.storage, &config)?;
+
+    let vote_module_msg = msg
+        .voting_module_instantiate_info
+        .into_wasm_msg(env.contract.address.clone());
+    let vote_module_msg: SubMsg<Empty> =
+        SubMsg::reply_on_success(vote_module_msg, VOTE_MODULE_INSTANTIATE_REPLY_ID);
+
+    let gov_module_msgs: Vec<SubMsg<Empty>> = msg
+        .governance_modules_instantiate_info
+        .into_iter()
+        .map(|info| info.into_wasm_msg(env.contract.address.clone()))
+        .map(|wasm| SubMsg::reply_on_success(wasm, GOV_MODULE_REPLY_ID))
+        .collect();
+    if gov_module_msgs.is_empty() {
+        return Err(ContractError::NoGovernanceModule {});
+    }
+
+    GOVERNANCE_MODULE_COUNT.save(deps.storage, &(gov_module_msgs.len() as u64))?;
+
+    Ok(Response::new()
+        .add_attribute("action", "instantiate")
+        .add_attribute("sender", info.sender)
+        .add_submessage(vote_module_msg)
+        .add_submessages(gov_module_msgs))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::ExecuteProposalHook { .. } => {
+            if !GOVERNANCE_MODULES.has(deps.storage, info.sender.clone()) {
+                return Err(ContractError::Unauthorized {});
+            }
+        }
+        _ => {
+            if info.sender != env.contract.address {
```

Comment:
instead of pattern matching, using if-else clauses could make this code much simpler to read.

Reply:
Agree. I'm just going to move this validation into the execute methods themselves. At one point it was a lot simpler to do it like this, but now this code is a little unwieldy.